archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,281 @@
+  { intro n, split_ifs,
+    { simp only [one_div, inv_nonneg, nat.cast_nonneg] },
+    { exact rfl.ge } },
+
+  rw ← summable_iff_not_tendsto_nat_at_top_of_nonneg hf at h,
+  rw summable_iff_vanishing at h,
+  specialize h (set.Ioo (-1 : ℝ) ((1 : ℝ) / 2)) (mem_nhds_sets is_open_Ioo (by norm_num)),
+  obtain ⟨s, h⟩ := h,
+  obtain ⟨k, hk⟩ := exists_nat_subset_range s,
+  use k,
+  intro x,
+
+  set P := {p ∈ range (x + 1) | p > k ∧ nat.prime p } with hP₁,
+  have hP₂ : P = filter (λ (p : ℕ), p > k ∧ nat.prime p) (range (x + 1)),
+  { rw [hP₁, sep_def], finish },
+
+  specialize h (filter (λ (n : ℕ), n > k) (range (x + 1))) _,
+  { rw disjoint_iff_ne,
+    intros a ha b hb,
+    rw mem_filter at ha,
+    obtain ⟨-, hak⟩ := ha,
+    exact ne_of_gt (lt_trans (mem_range.mp (hk hb)) hak) },
+  rw [← sum_filter, filter_filter, ← hP₂, set.mem_Ioo] at h,
+
+  exact h.right,
+end
+
+lemma lemma2_range_x_sdiff_M_eq_U {x k : ℕ} :
+  range x \ {e ∈ range x | ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k } =
+  finset.bUnion {p ∈ range (x + 1) | p > k ∧ nat.prime p } (λ p, {e ∈ range x | p ∣ (e + 1) }) :=
+begin
+  ext e,
+  simp only [mem_bUnion, not_and, mem_sdiff, sep_def, mem_filter, mem_range],
+  push_neg,
+  split,
+  { rintros ⟨hex, hexh⟩,
+    obtain ⟨p, ⟨hpp, hpe1⟩, hpk⟩ := hexh hex,
+    use p,
+    split,
+    { simp only [mem_filter, mem_range],
+    refine ⟨_, hpk, hpp⟩,
+    calc p ≤ e + 1 : (nat.le_of_dvd (nat.succ_pos e)) hpe1
+    ...    < x + 1 : nat.succ_lt_succ hex },
+    { exact ⟨hex, hpe1⟩ } },
+  { rintros ⟨p, hpfilter, ⟨hex, hpe1⟩⟩,
+    simp only [mem_filter, mem_range] at hpfilter,
+    obtain ⟨-, hpk, hpp⟩ := hpfilter,
+    rw imp_iff_right hex,
+    exact ⟨hex, ⟨p, ⟨hpp, hpe1⟩, hpk⟩⟩ },
+end
+
+lemma lemma3_card_U_le_x_mul_sum {x k : ℕ} :
+  (card
+    (finset.bUnion {p ∈ range (x + 1) | p > k ∧ nat.prime p } (λ p, {e ∈ range x | p ∣ (e + 1) }))
+  : ℝ)
+  ≤ x * (∑ p in {p ∈ range (x + 1) | p > k ∧ nat.prime p }, 1 / p) :=
+begin
+  let P := {p ∈ range (x + 1) | p > k ∧ nat.prime p },
+  let N := (λ p, {e ∈ range x | p ∣ (e + 1) }),
+  have h : card (finset.bUnion P N) ≤ ∑ p in P, card (N p) := card_bUnion_le,
+
+  calc  (card (finset.bUnion P N) : ℝ)
+      ≤ ∑ p in P, card (N p)  : by assumption_mod_cast
+  ... ≤ ∑ p in P, x * (1 / p) : by { refine sum_le_sum _,
+                                     intro p,
+                                     simp only [P, sep_def, mem_filter, mem_range],
+                                     rintros ⟨-, -, hpp⟩,
+                                     simp only [N, card_le_div_real (nat.prime.pos hpp)] }
+  ... = x * (∑ p in P, 1 / p) : mul_sum.symm,
+end
+
+lemma lemma4_aux_card_M1_le_2_pow_k {x k : ℕ} :
+  card {e ∈ range x | squarefree (e + 1) ∧ ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k} ≤ 2 ^ k :=
+begin
+  let M₁ := {e ∈ range x | squarefree (e + 1) ∧ ∀ p : ℕ, (nat.prime p ∧ p ∣ (e + 1)) → p ≤ k},
+  set f : finset ℕ → ℕ := λ s, (finset.prod s (λ a, a)) - 1 with hf_def,
```

Comment:
I would change this to
```suggestion
  let f := λ s, finset.prod s (λ a, a) - 1,
```
and then replace the `rw hf_def` below with `simp_rw f`